# 데이콘 Basic Summer
## 쇼핑몰 리뷰 평점 분류 경진대회 

## 두번째 Baseline

안녕하세요 데이커 여러분!  
데이콘 Basic Summer에 오신 것을 환영합니다.

올해도 무더운 여름이 찾아왔습니다.  
하지만 여름이라고 하면 역시 Shopping 🛒을 빼놓을 순 없겠죠! 

<img src='https://user-images.githubusercontent.com/92353013/177723710-5b62e89f-f46b-4575-b356-f2899e288017.png' align=center>

데이콘과 함께, 초보부터 고수까지 모두가 참여하는 데이콘 Basic Summer   
**쇼핑몰 리뷰 평점 분류 대회**를 소개합니다.

이번 대회는 쇼핑몰 리뷰 데이터를 통해 평점을 분류하는 대회입니다.

리뷰를 보고 1, 2, 4, 5점을 분류하는 간단한 대회 처럼 보이지만,   
자연어 처리를 처음 접하는 분들에게는 어디부터 시작해야 하는지 막막하실 겁니다.

베이스라인을 통해 자연어 처리 기초에 입문하여 멋진 데이커🥇가 되어보세요.

이번 두번째 베이스라인에서는 **TF-IDF를 활용**해 문장 벡터를 만들기 위한

**TfidfVectorizer 을 적용**하여 모델의 성능을 높여보겠습니다.

베이스라인을 통해 자연어 처리 기초에 입문해보세요!

<p style='text-align: right;'> July, 2022 | All rights reserved by Dacon</p>

# I. 데이터 살펴보기

데이터 분석을 하기 위해 가장 먼저 해야할 일이 무엇일까요?

바로 데이터를 살펴보는 일입니다.  
주어진 데이터를 살펴보며 데이터의 모양과 대략적인 내용을 파악하고, 무슨 분석이 가능한지 확인하여 분석의 방향을 잡는 과정이 필요합니다.

대회 페이지의 [데이터] 탭에서 압축 파일을 다운받아 압축을 풀어주었다면 현재 폴더 구조는 아래와 같을 것입니다.

(현재 위치)  
- baseline.ipynb
- dataset (폴더)
  - train.csv
  - test.csv
  - sample_submission.csv

이번 대회의 데이터는 총 3가지이며 모두 csv 파일입니다.
1. train.csv : 데이터 분석 모델을 학습(train)하기 위해 사용하는 데이터
2. test.csv : 데이터 분석 모델을 통해 정답을 예측하기 위해 사용하는 데이터
3. sample_submission.csv : 예측한 정답 값을 기록하여 제출하기 위한 샘플 정답 파일

## 1. 데이터 준비

먼저 분석하려는 데이터를 작업장으로 가져오는 작업이 필요합니다.

이를 위해서 파이썬 라이브러리 중 하나인 Pandas를 이용합니다.

pandas 라이브러리는 엑셀과 같은 행과 열로 이루어진 테이블(table) 형식의 데이터를 편하게 다루도록 해주는 라이브러리입니다.

In [1]:
#csv 형식의 training 데이터를 로드합니다.
import pandas as pd 

train = pd.read_csv('dataset/train.csv')
test = pd.read_csv('dataset/test.csv')

#데이터 살펴보기 위해 데이터 최상단의 5줄을 표시합니다.
train.head() 

id                                            reviews  target
0   0                                     조아요 처음구입 싸게햇어요       2
1   1  생각보다 잘 안돼요 매지 바른지 하루밖에 안됐는데ㅠㅠ 25천원가량 주고 사기 너무 ...       1
2   2  디자인은괜찮은데 상품이 금이가서 교환했는데 두번째받은상품도 까져있고 안쪽에 금이가져...       2
3   3  기전에 이 제품말고 이마트 트레이더스에서만 팔던 프리미엄 제품을 사용했었습니다. 샘...       2
4   4                                  튼튼하고 손목을 잘 받쳐주네요~       5

## 2. 데이터 확인

이제 데이터가 어떻게 생겼는지 살펴봅시다.

처음 데이터를 살펴 볼 때에는 pandas의 **head()**, **tail()** 그리고 **info()** 메소드를 활용하여 데이터를 살펴보는 경우가 많습니다.

이는 데이터와 처음과 끝을 살펴봄으로서, **이상치나 결측치**와 같은 데이터를 살펴보기 위함입니다.

In [2]:
# 데이터를 살펴보기 위해 데이터 최상단의 5줄을 표시합니다.
train.head()

id                                            reviews  target
0   0                                     조아요 처음구입 싸게햇어요       2
1   1  생각보다 잘 안돼요 매지 바른지 하루밖에 안됐는데ㅠㅠ 25천원가량 주고 사기 너무 ...       1
2   2  디자인은괜찮은데 상품이 금이가서 교환했는데 두번째받은상품도 까져있고 안쪽에 금이가져...       2
3   3  기전에 이 제품말고 이마트 트레이더스에서만 팔던 프리미엄 제품을 사용했었습니다. 샘...       2
4   4                                  튼튼하고 손목을 잘 받쳐주네요~       5

In [3]:
# 데이터를 살펴보기 위해 데이터 최하단의 5줄을 표시합니다.
train.tail()

id                                      reviews  target
24995  24995          일단향이너무너무좋은데지속력만좋다면좋겠지만..워낙저렴해그래도좋아여       5
24996  24996  동일한 업체에서 발송하고 한 상자로 배송되었는데 택배비는 이중 처리되었습니다.       2
24997  24997                      향이 좋아요!! 다른 향도 시켜보고 싶어요       4
24998  24998    펼쳐보니 끝부분쯤에 50cm정도 찟어져 있습니다 날짜관계로 구냥 사용합니다       2
24999  24999                             호호애미 제품은 늘 믿고 써요       5

In [4]:
# 데이터 결측치 및 변수들의 타입 확인
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       25000 non-null  int64 
 1   reviews  25000 non-null  object
 2   target   25000 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 586.1+ KB


info() 메소드를 이용해 데이터 결측치 및 변수들의 타입을 확인해본 결과,           
타겟 변수를 포함하여 총 3개의 변수와 25000 행으로 이루어져 있습니다.     

Non-Null Count 부분을 보니 모든 행에서 결측치(NA)는 발견되지 않았네요!

# II. 데이터 전처리

## 1. Feature Selection

모델링을 하기 앞서, 독립변수 X와 종속변수(타겟변수) y를 설정해야 합니다.  

우리의 목적은 주어진 쇼핑몰 리뷰 데이터(독립변수)를 이용하여 평점(target)을 예측하는 것입니다.

그럼 쇼핑몰 리뷰가 X, 평점(target)이 y가 되겠죠?

In [5]:
X = train.reviews #training 데이터에서 피쳐 추출
y = train.target #training 데이터에서 타겟 추출

In [6]:
X.head() #데이터 살펴보기

0                                       조아요 처음구입 싸게햇어요
1    생각보다 잘 안돼요 매지 바른지 하루밖에 안됐는데ㅠㅠ 25천원가량 주고 사기 너무 ...
2    디자인은괜찮은데 상품이 금이가서 교환했는데 두번째받은상품도 까져있고 안쪽에 금이가져...
3    기전에 이 제품말고 이마트 트레이더스에서만 팔던 프리미엄 제품을 사용했었습니다. 샘...
4                                    튼튼하고 손목을 잘 받쳐주네요~
Name: reviews, dtype: object

In [7]:
y.head() #데이터 살펴보기

0    2
1    1
2    2
3    2
4    5
Name: target, dtype: int64

위의 데이터프레임과 같이 text가 깔끔히 정리되었습니다!

## 2. 토큰화

토큰화(Tokenization)란 말뭉치(Corpus)를 주어진 단위(Token)로 나누는 과정을 의미합니다.

단어 단위로 나누는 단어 토큰화(word tokenization)의 예시를 살펴보겠습니다.

In [8]:
text = "Dacon 베이직 서머에서, 시원하게 데이터 분석 하자 !" # 예시 문장을 정의합니다.

한국어에 대한 전처리는 대표적으로 okt 패키지를 사용합니다. 

In [9]:
from konlpy.tag import Okt
okt=Okt()  

print('단어 토큰화 결과 ==>', okt.morphs(text))
print('명사 추출 결과 ==>', okt.nouns(text)) 

단어 토큰화 결과 ==> ['Dacon', '베이직', '서머', '에서', ',', '시원하게', '데이터', '분석', '하자', '!']
명사 추출 결과 ==> ['베이직', '서머', '데이터', '분석', '하자']


단어 단위로 깔끔하게 분리되네요. "!", "." 등의 구두점도 하나의 단어로 취급되는 것을 확인할 수 있습니다.

이렇게 토큰화에 대해서 간단히 살펴보았습니다.

그럼 쇼핑몰 리뷰 데이터에서도 같은 방법으로 토큰화를 진행시켜  
'tokenized_stem' 열에 새로 토큰화 된 데이터를 생성해보겠습니다.

In [10]:
tokenized = [] # 데이터프레임의 한 컬럼으로 추가할 리스트
def token(dataframe):
    for sentence in dataframe['reviews']: # 전처리된 리뷰들을 하나씩 꺼내옵니다
        tokens = okt.morphs(sentence)
        tokenize = " ".join(tokens) # tokens라는 리스트 안의 형태소들을 띄어쓰기로 분리된 하나의 문자열로 join시켜줍니다.
        tokenized.append(tokenize) # 형태소 단위로 띄어쓰기된 문자열을 최종 리스트에 추가해줍니다
        dataframe["tokenized_stem"] = pd.DataFrame(tokenized) # 리스트를 데이터프레임으로 변환해 tokenized_stem라는 컬럼명으로 추가해줍니다.
    
token(train)
token(test)

train.head() # 데이터 확인

id                                            reviews  target  \
0   0                                     조아요 처음구입 싸게햇어요       2   
1   1  생각보다 잘 안돼요 매지 바른지 하루밖에 안됐는데ㅠㅠ 25천원가량 주고 사기 너무 ...       1   
2   2  디자인은괜찮은데 상품이 금이가서 교환했는데 두번째받은상품도 까져있고 안쪽에 금이가져...       2   
3   3  기전에 이 제품말고 이마트 트레이더스에서만 팔던 프리미엄 제품을 사용했었습니다. 샘...       2   
4   4                                  튼튼하고 손목을 잘 받쳐주네요~       5   

                                      tokenized_stem  
0                                 조아요 처음 구입 싸게 햇 어 요  
1  생각 보다 잘 안 돼요 매지 바른지 하루 밖에 안 됐는데 ㅠㅠ 25천원 가량 주고 ...  
2  디자인 은 괜찮은데 상품 이 금 이 가서 교환 했는데 두번째 받은 상품 도 까져있고...  
3  기전 에 이 제품 말고 이마트 트 레이더스 에서만 팔던 프리미엄 제품 을 사용 했었...  
4                                튼튼하고 손목 을 잘 받쳐주네요 ~

## 3. 품사 태깅(POS Tagging)

자 이제 다음 단계인 품사 태깅에 대해 알아봅시다.

품사 태깅이란 주어진 텍스트를 형태소 단위로 나눈 뒤, 각 형태소에 해당 품사를 태깅하여 리스트화 하는 것입니다.

그럼 토큰화된 단어의 품사를 nltk의 pos_tag메소드를 이용하여 분리해보도록 하겠습니다.

In [11]:
# okt.pos() 를 이용
print('품사 태깅 결과 ==>', okt.pos(text))  

품사 태깅 결과 ==> [('Dacon', 'Alpha'), ('베이직', 'Noun'), ('서머', 'Noun'), ('에서', 'Josa'), (',', 'Punctuation'), ('시원하게', 'Adjective'), ('데이터', 'Noun'), ('분석', 'Noun'), ('하자', 'Noun'), ('!', 'Punctuation')]


품사 태깅은 꼭 필요한 품사(ex. 명사, 동사)를 추출할 때 유용하게 쓰일 수 있습니다!

뉴스 텍스트는 명사만으로 그룹 분류를 판단하기 어렵기 때문에

우선 임의로 명사, 동사, 형용사, 부사를 추출하여 사용해보겠습니다. (판단에 따라 설정해주세요 !)

In [12]:
def postagging(dataframe):
    main_pos = [] # 데이터프레임의 새 컬럼이 될 리스트
    for sentence in dataframe['tokenized_stem']: # 리뷰들을 하나씩 가져옵니다
        pos = okt.pos(sentence) # 형태소 분석을 진행하고 해당 리스트를 pos라는 변수로 받습니다
        main_words = [word_pos[0] for word_pos in pos if word_pos[1] in ('Adjective', #형용사
                                                                         'Noun', #명사
                                                                         'Adjective', #부사
                                                                         'Verb', #동사
                                                                         'Number',
                                                                         'KoreanParticle'
                                                                         )] # 가져오고자 하는 품사에 해당하면 해당 형태소를 main_words 리스트에 추가합니다.
        main_words_str = " ".join(main_words) # main_words 리스트 안의 형태소들을 띄어쓰기로 분리된 하나의 문자열로 join시켜줍니다.
        main_pos.append(main_words_str) # 선택한 형태소들로 이루어진 문자열을 최종 리스트에 추가해줍니다
        dataframe["main_pos"] = pd.DataFrame(main_pos) # 리스트를 데이터프레임으로 변환해 main_pos라는 컬럼명으로 추가해줍니다.

postagging(train)
#postagging(test) ## test셋도 똑같이 품사태깅을 적용해줍니다.
train.head() # 데이터 확인

id                                            reviews  target  \
0   0                                     조아요 처음구입 싸게햇어요       2   
1   1  생각보다 잘 안돼요 매지 바른지 하루밖에 안됐는데ㅠㅠ 25천원가량 주고 사기 너무 ...       1   
2   2  디자인은괜찮은데 상품이 금이가서 교환했는데 두번째받은상품도 까져있고 안쪽에 금이가져...       2   
3   3  기전에 이 제품말고 이마트 트레이더스에서만 팔던 프리미엄 제품을 사용했었습니다. 샘...       2   
4   4                                  튼튼하고 손목을 잘 받쳐주네요~       5   

                                      tokenized_stem  \
0                                 조아요 처음 구입 싸게 햇 어 요   
1  생각 보다 잘 안 돼요 매지 바른지 하루 밖에 안 됐는데 ㅠㅠ 25천원 가량 주고 ...   
2  디자인 은 괜찮은데 상품 이 금 이 가서 교환 했는데 두번째 받은 상품 도 까져있고...   
3  기전 에 이 제품 말고 이마트 트 레이더스 에서만 팔던 프리미엄 제품 을 사용 했었...   
4                                튼튼하고 손목 을 잘 받쳐주네요 ~   

                                            main_pos  
0                                   조아요 처음 구입 싸게 햇 요  
1  생각 보다 잘 안 돼요 매지 바른지 하루 안 됐는데 ㅠㅠ 25천원 가량 주고 사기 ...  
2  디자인 은 괜찮은데 상품 이 금 이 가서 교환 했는데 두번째 받은 상품 도 까져있고...  
3  기전 이 제품 이마트 트 레이더스 팔던 프리미엄 제품 사용 했었습니다 샘플 써 보고...  
4                                    튼튼하고 손목 잘 받쳐주네요

In [27]:
postagging(test)

자 이제 총 두 개의 전처리된 컬럼이 생성되었습니다!

어떤 전처리 과정이 성능이 가장 잘 나올지는 미지수입니다. 직접 다양한 시도를 통해 최적의 전처리 프로세스를 찾아보세요!

이번 베이스라인에서는 마지막에 생성한 main_pos 컬럼을 이용해 모델 학습을 진행하겠습니다.



# III. 모델링

## 1. train/vaildation 정의

이제 우리는 모델을 만들어야 하는데,   
모델을 만들기 앞서 내가 만들 모델이 유의미한지 알아야 좋은 결론을 도출 할 수 있겠죠?   

하지만 단순히 주어진 데이터로 train셋 데이터를 학습시켜서 목표값인 test 셋에 집어 넣어 예측값을 도출하면,   
모델이 적합한지 알 방법이 없습니다.   


따라서 train 셋에서 70%를 학습을 시키는데 쓰고, 나머지 25%를 모델이 유의미한지 검증하는데 사용할 것입니다.   
이후 유의미하다면, 실제로 제출할 모델을 만들 때, train 셋 100%를 학습을 시키면 되겠죠?

자 그럼 train 셋에서 학습시킬 셋 train_data와 검증할 셋 val_data를 만들어 봅시다!

In [13]:
from sklearn.model_selection import train_test_split

data = train.copy()
train, val = train_test_split(data)
train.reset_index(inplace=True) # 전처리 과정에서 데이터가 뒤섞이지 않도록 인덱스를 초기화해주었습니다.
val.reset_index(inplace=True)

In [14]:
print( 'train 데이터 셋 모양 :', train.shape)
print( 'val 데이터 셋 모양 :', val.shape)

train 데이터 셋 모양 : (18750, 6)
val 데이터 셋 모양 : (6250, 6)


train 셋은 18750개, val 셋은 6250개 데이터로 나뉘어 진 것을 확인할 수 있습니다.

## 2. 벡터화

CountVectorizer를 사용하였습니다.

이 부분에 대한 설명은 [첫번째 베이스라인](https://dacon.io/competitions/official/235938/codeshare/5365?page=1&dtype=recent)에 자세히 나와있으니 참고 바랍니다.

In [15]:
X_train = train.main_pos #training 데이터에서 문서 추출
y_train = train.target #training 데이터에서 라벨 추출

In [16]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer() #countvectorizer 생성
vectorizer.fit(X_train) # countvectorizer 학습
X_train_vec = vectorizer.transform(X_train) # transform

## 3. 모델 학습

In [17]:
from sklearn.linear_model import LogisticRegression #모델 불러오기
model = LogisticRegression(max_iter=500) #객체에 모델 할당
model.fit(X_train_vec, y_train) #모델 학습

LogisticRegression(max_iter=500)

## 4. 검증셋으로 모델 성능 검증

트레인 셋에서 학습된 모델을   
검증 셋을 통해서 얼마나 성능이 나오는지 확인해보도록 하겠습니다.

In [18]:
X_val = val.main_pos #validation 데이터에서 전처리된 문서 추출
y_val = val.target #validation 데이터에서 라벨 추출

X_val_vec = vectorizer.transform(X_val) # train셋으로 fit한 벡터라이저 이용해 transform

In [19]:
y_pred = model.predict(X_val_vec)
print(y_pred)

[2 2 2 ... 5 5 2]


In [20]:
from sklearn import metrics
print('Logistic Regression 의 예측 정확도는', round(metrics.accuracy_score(y_val, y_pred),3)) # 정확도 확인

Logistic Regression 의 예측 정확도는 0.624


검증 셋을 예측해본 결과 61.2%의 정확도를 얻었습니다!

조금 더 성능을 끌어 올려보는 작업을 해보겠습니다.

# 5. TF-IDF

Tf-Idf (Term Frequency - Inverse Document Frequency)는 

TF(단어 빈도, term frequency) : 특정한 단어가 문서 내에 얼마나 자주 등장하는지를 나타내는 값  
IDF(역문서 빈도, inverse document frequency) : 단어 자체가 문서군 내에서 자주 사용하여 얼마나 그 단어가 흔하게 등장하는지를 나타내는 값의 역수   
를 곱한 값입니다.

즉, 특정 문서 내에서 단어 빈도가 높을 수록, 그리고 전체 문서들 중 그 단어를 포함한 문서가 적을 수록 TF-IDF값이 높아집니다.

특정 단어를 포함하는 문서들이 많을 수록 로그 함수 안의 값이 1에 가까워지게 되고,   
이 경우 IDF값과 TF-IDF값은 0에 가까워지게 되는 것입니다.

따라서 이 값을 이용하면 모든 문서에 흔하게 나타나는 단어를 걸러내는 효과를 얻을 수 있습니다. 

TF-IDF는 주로 문서의 유사도를 구하는 작업, 검색 시스템에서 검색 결과의 중요도를 정하는 작업,  
 문서 내에서 특정 단어의 중요도를 구하는 작업 등에 쓰입니다.

그럼, TF-IDF를 활용해 문장 벡터를 만들기 위한   
TfidfVectorizer 을 적용하여 모델의 성능을 높여보겠습니다!

In [21]:
from sklearn.feature_extraction.text import TfidfVectorizer

X_train = train.main_pos #training 데이터에서 문서 추출
y_train = train.target #training 데이터에서 라벨 추출

# TF-IDF Vectorization 적용하여 학습 데이터셋과 테스트 데이터 셋 변환. 
tfidf_vect = TfidfVectorizer()
tfidf_vect.fit(X_train)

X_val = val.main_pos #validation 데이터에서 전처리된 문서 추출
y_val = val.target #validation 데이터에서 라벨 추출

X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_val_tfidf_vect = tfidf_vect.transform(X_val) # train셋으로 fit한 벡터라이저 이용해 transform
print('학습 & 테스트 데이터 Text의 TfidfVectorizer Shape:',X_train_tfidf_vect.shape, X_val_tfidf_vect.shape)

# LogisticRegression을 이용하여 학습/예측/평가 수행. 
lr_clf = LogisticRegression(solver='liblinear') 
lr_clf.fit(X_train_tfidf_vect , y_train)
pred = lr_clf.predict(X_val_tfidf_vect)
print('TF-IDF Logistic Regression 의 예측 정확도는 {0:.3f}'.format(metrics.accuracy_score(y_val ,pred)))

학습 & 테스트 데이터 Text의 TfidfVectorizer Shape: (18750, 27238) (6250, 27238)
TF-IDF Logistic Regression 의 예측 정확도는 0.647


TF-IDF를 적용하여 검증 셋을 예측해본 결과 64.1%의 정확도로 정확도가 상승한 것을 알 수 있습니다 !



## 5. test.csv 분류하기


이제 본격적으로 대회에서 주어진 정답이 없는 test 데이터의 라벨을 예측해보겠습니다.

그럼, 트레인 셋과 검증 셋으로 나누어 성능을 확인해 준 것을 하나의 트레인 셋으로 통합하여

다시 모델을 만들어주겠습니다.

In [22]:
X_train = data.main_pos #전체 training 데이터에서 문서 추출
y_train = data.target #전체 training 데이터에서 라벨 추출

# TF-IDF Vectorization 적용하여 학습 데이터셋과 테스트 데이터 셋 변환. 
tfidf_vect = TfidfVectorizer(ngram_range=(1,2), max_df=300)
tfidf_vect.fit(X_train)

X_train_tfidf_vect = tfidf_vect.transform(X_train)

# LogisticRegression을 이용하여 학습/예측/평가 수행. 
lr_clf = LogisticRegression(solver='liblinear', C = 10) 
lr_clf.fit(X_train_tfidf_vect , y_train)

LogisticRegression(C=10, solver='liblinear')

이제 test셋을 모델에 넣어 예측값을 만들어 줍니다!

In [26]:
train

index     id                                            reviews  \
0      22210  22210  맛이 더럽게 없고 물탄거같음 반은 물러서 버리고 터져서왔고 나머지 안터진것마저 맛이...   
1      15826  15826  성분빼고 딱히 만족할만한 점이 없어요 보습도 잘안되고 냄새도 별로고 근데 오일을 섞...   
2      23053  23053           가운데 다리를 끼웠더니 다리높이들이 안맞고 움직여요 그래서 빼 놓았어요~   
3       1086   1086  핏이 좋아서 가성비 좋은 상품같아요~ 다만 얇아서 쉽게 뜯어지진 않을까 걱정은 되지...   
4      24468  24468                 소음만크고,때는잘안밀립니다. 쓸데없는돈을쓴것같아.무지아깝습니다   
...      ...    ...                                                ...   
18745  18908  18908                      오랜만에 시켜먹는 씨알상사~~ 여전히 맛 좋아요 :)   
18746  23524  23524                  이거 중고아닌가싶음... 의심은가지만 귀찮아서 그냥대충쓰겠음   
18747  18185  18185                                저는 백산수만 마셔요 물맛이 좋아요   
18748  11218  11218                                     좋습니다. 가격도 착하네요   
18749  18197  18197  한번입고 못입게생겼어요. 다 늘어지고 후질근해지고. 허리는 왜이리 큰지...돈 아까...   

       target                                     tokenized_stem  \
0           1  맛 이 더럽게 없고 물탄 거 같음 반 은 물러서 버리고 터져서 왔고 나머지 안 터진...   
1           2  성분 빼고 딱히 만족할만 한 점 이 없어요 보습 도 잘 안되고 냄새 도 별로 고 근...   
2           2     가운데 다리 를 끼웠더니 다리 높이 들 이 안 맞고 움직여요 그래서 빼 놓았어요 ~   
3           5  핏 이 좋아서 가성 비 좋은 상품 같아요 ~ 다만 얇아서 쉽게 뜯어지진 않을까 걱정...   
4           2  소음 만 크고 , 때 는 잘 안 밀립니다 . 쓸데없는 돈 을 쓴것 같아 . 무지 아...   
...       ...                                                ...   
18745       5                 오랜 만 에 시켜 먹는 씨알 상사 ~~ 여전히 맛 좋아요 :)   
18746       1          이 거 중고 아닌가싶음 ... 의심 은 가지 만 귀찮아서 그냥 대충 쓰겠음   
18747       5                            저 는 백 산수 만 마셔요 물맛 이 좋아요   
18748       5                                   좋습니다 . 가격 도 착하네요   
18749       1  한번 입고 못 입게 생겼어요 . 다 늘어지고 후질 근 해지 고 . 허리 는 왜 이리...   

                                                main_pos  
0      맛 이 더럽게 없고 물탄 거 같음 반 은 물러서 버리고 터져서 왔고 나머지 안 터진...  
1      성분 빼고 딱히 만족할만 한 점 이 없어요 보습 도 잘 안되고 냄새 도 별로 고 오...  
2               가운데 다리 를 끼웠더니 다리 높이 들 이 안 맞고 움직여요 빼 놓았어요  
3      핏 이 좋아서 가성 비 좋은 상품 같아요 다만 얇아서 쉽게 뜯어지진 않을까 걱정 은...  
4             소음 만 크고 때 는 잘 안 밀립니다 쓸데없는 돈 쓴것 같아 무지 아깝습니다  
...                                                  ...  
18745                            만 시켜 먹는 씨알 상사 여전히 맛 좋아요  
18746              이 거 중고 아닌가싶음 의심 은 가지 만 귀찮아서 그냥 대충 쓰겠음  
18747                            저 는 백 산수 만 마셔요 물맛 이 좋아요  
18748                                     좋습니다 가격 도 착하네요  
18749    한번 입고 못 입게 생겼어요 늘어지고 후질 해지 고 허리 는 왜 큰지 돈 아까워요 ㅠ  

[18750 rows x 6 columns]

In [28]:
X_test = test.main_pos
X_test_vec = tfidf_vect.transform(X_test)
pred_test = lr_clf.predict(X_test_vec)
print(pred_test)

[2 1 2 ... 4 2 5]


# IV. 데이콘 제출하기

## 1. submission 파일 생성

이제 예측한 값을 **sample_submission.csv**에 넣어 제출용 파일을 생성해봅시다.            

제출한 뒤 리더보드를 통해 결과를 확인합시다.

In [32]:
# 제출용 sample 파일을 불러옵니다.
submission = pd.read_csv('dataset/sample_submission.csv')
submission.head()

id  target
0   0       0
1   1       0
2   2       0
3   3       0
4   4       0

In [33]:
# 위에서 구한 예측값을 그대로 넣어줍니다.
submission['target'] = pred_test

# 데이터가 잘 들어갔는지 확인합니다.
submission

id  target
0          0       2
1          1       1
2          2       2
3          3       2
4          4       5
...      ...     ...
24995  24995       5
24996  24996       2
24997  24997       4
24998  24998       2
24999  24999       5

[25000 rows x 2 columns]

submission을 csv 파일로 저장합니다.   
index=False란 추가적인 id를 부여할 필요가 없다는 뜻입니다.    
정확한 채점을 위해 꼭 index=False를 넣어주세요.

In [ ]:
submission.to_csv("submission_baseline2.csv", index=False)

## 2. 데이콘 페이지 제출
다음으로, 실제 데이콘 홈페이지에서 제출해보도록 하겠습니다.

이 과정을 통해 우리의 모델이 목표로 하는 데이터에 대해 어느 정도의 성능을 가지는지, 다른 사람들은 어느 정도의 성능을 얻었는지 확인할 수 있습니다.

<img src='https://user-images.githubusercontent.com/92353013/177723966-77b77340-c0f6-4873-95fc-9eb8c3812700.png'>

1. 먼저 대회 페이지에서 제출 탭의 버튼을 누르면 파일 업로드 화면으로 전환됩니다.

2. 전환된 화면에서 서류모양 아이콘을 클릭하여 제출할 파일을 선택하거나 파일을 마우스로 끌어와 서류모양 아이콘 위에 던져두면 파일을 업로드 할 수 있습니다.

3. 업로드가 완료되었다면 까만색 제출 버튼을 클릭합니다.

# V. 마치며

<img src='https://user-images.githubusercontent.com/92353013/177723980-8d4c4662-9ba7-4834-9827-0b88824480fb.png'>

축하합니다! 데이콘 베이직 자연어 처리를 완료하셨습니다!  

이번 두번째 Baseline 통해 어떤 방식으로 분석을 진행해야겠다는 인사이트를 얻으셨으면 좋겠습니다.

여러분도 여러분만의 분석을 진행하시고 **코드공유**에 올려주세요!

데이콘 스쿨 장학금, 스타벅스 기프티콘, 데이콘 굿즈 등 특별한 선물이 기다립니다!